pandas/_libs/algos.pyx

```diff
@@ -797,212 +797,202 @@ def rank_1d(
+                    out[_as[j]] = i - grp_start - dups + 2
+            elif tiebreak == TIEBREAK_MAX:
+                for j in range(i - dups + 1, i + 1):
+                    out[_as[j]] = i - grp_start + 1
+            elif tiebreak == TIEBREAK_FIRST:
+                for j in range(i - dups + 1, i + 1):
+                    if ascending:
+                        out[_as[j]] = j + 1 - grp_start
                     else:
-                        for j in range(i - dups + 1, i + 1):
-                            ranks[argsorted[j]] = j + 1
-                elif tiebreak == TIEBREAK_FIRST_DESCENDING:
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = 2 * i - j - dups + 2
-                elif tiebreak == TIEBREAK_DENSE:
-                    total_tie_count += 1
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = total_tie_count
-                sum_ranks = dups = 0
-
-    else:
-        with nogil:
-            # TODO: why does the 2d version not have a nogil block?
-            for i in range(n):
-                sum_ranks += i + 1
-                dups += 1
-
-                val = sorted_data[i]
-
-                if rank_t is not uint64_t:
-                    isnan = sorted_mask[i]
-                    if isnan and keep_na:
-                        ranks[argsorted[i]] = NaN
-                        continue
-
-                count += 1.0
-
-                if rank_t is object:
-                    condition = (
-                        i == n - 1 or
-                        are_diff(sorted_data[i + 1], val) or
-                        i == non_na_idx
-                    )
+                        out[_as[j]] = 2 * i - j - dups + 2 - grp_start
+            elif tiebreak == TIEBREAK_DENSE:
+                for j in range(i - dups + 1, i + 1):
+                    out[_as[j]] = grp_vals_seen
+
+            # look forward to the next value (using the sorting in _as)
+            # if the value does not equal the current value then we need to
+            # reset the dups and sum_ranks, knowing that a new value is
+            # coming up. the conditional also needs to handle nan equality
+            # and the end of iteration
+            if next_val_diff or (mask[_as[i]] ^ mask[_as[i+1]]):
+                dups = sum_ranks = 0
+                grp_vals_seen += 1
+                grp_tie_count += 1
+
+            # Similar to the previous conditional, check now if we are
+            # moving to a new group. If so, keep track of the index where
+            # the new group occurs, so the tiebreaker calculations can
+            # decrement that from their position. fill in the size of each
+            # group encountered (used by pct calculations later). also be
+            # sure to reset any of the items helping to calculate dups
+            if at_end or (check_labels and (labels_[_as[i]] != labels_[_as[i+1]])):
```

Comment:
The `check_labels` is not necessary for correctness since `labels_` is zeroed if no labels are passed, but short-circuiting to avoid the unnecessary check `labels_[_as[i]] != labels_[_as[i+1]]` seemed to give a small perf boost